### PR TITLE
Add Meter SDK Class

### PR DESCRIPTION
**Pull request by Brandon-Kimberly**
This PR adds the header and implementation files for the SDK Meter class as well as a suite of unit tests. This class functions as the accumulator in the [SDK spec](https://github.com/open-telemetry/opentelemetry-specification/blob/64bbb0c611d849b90916005d7714fa2a7132d0bf/specification/metrics/sdk.md#accumulator). This class provides the user a way to create new metric instruments, and collect those instruments’ updates for exporting as well as a facility for recording to a batch of instruments in a single API call.

**IMPORTANT NOTES:**

1. The code in this PR depends on the SDK metric instruments code in PR #179 and PR #191, and the Meter API code in PR #162.
2. THIS CODE DOES NOT CURRENTLY COMPILE. (<-- This issue has been resolved. See bottom of description for details)
    1. I will point out the specific code where I am seeing this problem in inline comments. I have also posted a description of this issue directly below and in [issue #216](https://github.com/open-telemetry/opentelemetry-cpp/issues/216).

**Issue**

The problem I am running into currently is that, I need to cast from an API component to an SDK component. This is because all functions and variables related to the collection of metrics are, according to the spec, only defined in the SDK. This means that we cannot have virtual functions in the API to access the information we need to collect and send for export such as the instruments’ aggregators.

The issue with that is, nostd::shared_ptr does not work with std::dynamic_pointer_cast. We cannot simply cast to a raw pointer as this could cause memory leaks or segfaults should we try to access memory that was freed by the user’s shared_ptr to the instrument. We need to use nostd::shared_ptr because the pointers to the created instruments that are returned to the user, must also be placed into a map so that their measurements can be collected later. The pointer in the map must be deleted when the user’s pointer goes out of scope or it will segfault on the next call to Collect(). I am unsure about how to implement this with raw pointers.

I believe the cleanest solution to this problem could be to implement dynamic_pointer_cast for nostd::shared_ptr. I have tested my class with std::shared_ptr and std::dynamic_pointer_cast and all functionality works as intended. In order to provide this functionality to nostd::shared_ptr I believe we would need to add an aliasing constructor to the class. I have looked into this a bit but I am out of my depth with STL code.

Any help or advice on how to fix this problem would be greatly appreciated. I am currently *blocked* on this problem and it is preventing me from being able to complete contributing a fully functional metrics API and SDK.

**UPDATE**
Thanks to help from @pyohannes I was able to resolve the issue involving dynamic_pointer_cast. This PR is still dependent on code in PR #191 and PR #162 and should not be merged before them but the code is complete and awaiting more reviews.

### CODE REVIEW

**Comment by Brandon-Kimberly** [comment]
This PR is ready to be reviewed again now that the issue with nostd::shared_ptr has been resolved.

**Comment by reyang** [comment]
@Brandon-Kimberly would you rebase and resolve the conflicts?

**Comment by Brandon-Kimberly** [comment]
@reyang This PR has been rebased and all CI checks pass. I believe it is ready to merge pending review. :)

File: sdk/include/opentelemetry/sdk/metrics/meter.h [hunk 1]
```diff
@@ -20,254 +25,360 @@ class Meter final : public metrics_api::Meter, public std::enable_shared_from_th
...
 
+  /**
+   * Creates a Counter with the passed characteristics and returns a shared_ptr to that Counter.
+   *
+   * @param name the name of the new Counter.
+   * @param description a brief description of what the Counter is used for.
+   * @param unit the unit of metric values following https://unitsofmeasure.org/ucum.html.
+   * @param enabled a boolean value that turns on or off the metric instrument.
+   * @return a shared pointer to the created Counter.
+   * @throws IllegalArgumentException if {@code name} is null
```
Comment (reyang) [inline sdk/include/opentelemetry/sdk/metrics/meter.h:36]: Is this copied from Java and is it valid here?

File: sdk/include/opentelemetry/sdk/metrics/meter.h [hunk 2]
```diff
@@ -20,254 +25,360 @@ class Meter final : public metrics_api::Meter, public std::enable_shared_from_th
...
   std::string library_name_;
   std::string library_version_;
+
+  std::mutex metrics_lock_;
+  std::mutex observers_lock_;
 };
 }  // namespace metrics
 }  // namespace sdk
-OPENTELEMETRY_END_NAMESPACE
+OPENTELEMETRY_END_NAMESPACE
```
Comment (reyang) [inline sdk/include/opentelemetry/sdk/metrics/meter.h:384]: nit, add the blank line back.

File: sdk/src/metrics/meter.cc [hunk 3]
```diff
@@ -0,0 +1,775 @@
...
+{
+namespace metrics
+{
+nostd::shared_ptr<metrics_api::Counter<short>> Meter::NewShortCounter(
+    nostd::string_view name,
+    nostd::string_view description,
+    nostd::string_view unit,
+    const bool enabled)
+{
+  if (!IsValidName(name) || NameAlreadyUsed(name))
```
Comment (reyang) [inline sdk/src/metrics/meter.cc:13]: `NameAlreadyUsed` is not thread safe, correct?
For example, even if this check passed, there could still be race condition before we actually call `short_metrics_.insert`:
1. Thread A checked NameAlreadyUsed("foo") and it returned false.
2. Thread B checked NameAlreadyUsed("foo") and it returned false.
3. Thread A inserted the newly created "foo".
4. Thread B inserted the newly created "foo".

File: sdk/include/opentelemetry/sdk/metrics/meter.h (hunk 1 above)
Comment (Brandon-Kimberly) [reply sdk/include/opentelemetry/sdk/metrics/meter.h]: Good catch! Yes, it is. I updated the comments to better reflect the actual code.

File: sdk/include/opentelemetry/sdk/metrics/meter.h (hunk 2 above)
Comment (Brandon-Kimberly) [reply sdk/include/opentelemetry/sdk/metrics/meter.h]: Fixed

File: sdk/src/metrics/meter.cc (hunk 3 above)
Comment (Brandon-Kimberly) [reply sdk/src/metrics/meter.cc:13]: Hmm that is a good point. Moving the mutex.locks() in the CreateInstrument functions to the top of the function definitions would resolve this issue, yes?

Comment (Brandon-Kimberly) [reply sdk/src/metrics/meter.cc:13]: Or, rather I think I may need to lock both the metrics_lock and observers_lock in the NameAlreadyUsed function. I believe the solution in my previous comment would still allow for a similar situation to occur if one instrument was synchronous and the other asynchronous.

Comment (Brandon-Kimberly) [reply sdk/src/metrics/meter.cc:13]: Ok, it required a tiny bit of refactoring but I believe that I have solved this issue now in an elegant way. Thanks for the feedback!

**Pull request merged**